lib/spack/spack/spec.py

```diff
@@ -4861,6 +4862,46 @@ def reconstruct_virtuals_on_edges(spec):
             edge.update_virtuals([vspec])
 
 
+def format_path(spec, format_string):
+    """Given a `format_string` that is intended as a path, generate a string
+    like from `Spec.format`, but eliminate extra path separators introduced by
+    formatting of Spec properties.
+
+    Path separators explicitly added to the string are preserved, so for example
+    "{name}/{version}" would generate a directory based on the Spec's name, and
+    a subdirectory based on its version; this function guarantees though that
+    the resulting string would only have two directories (i.e. that if under
+    normal circumstances that `str(Spec.version)` would contain a path
+    separator, it would not in this case).
+
+    Note that POSIX or Windows-style paths can be supplied on either POSIX or
+    Windows, and both "/" and "\" are treated as path separators (as a side
+    effect, "invalid" paths which mix these will be converted to OS-appropriate
+    paths with consistent separators).
+    """
+    # A root on Linux ("/"), or part of a Windows device namespace ("\")
+    any_sep = r"[/\\]"
+    # A drive on Windows (e.g. "C:\\")
+    drive = r"[a-zA-Z]:[/\\]"
+    root_pattern = rf"^({any_sep}|{drive})"
+    if re.match(root_pattern, format_string):
+        # Note: we could attempt to construct the absolute formatted path (i.e.
+        # produce a valid path instead of failing here), but we would want to
+        # account for the following cases:
+        # "\\root\{name}\{version}", "C:\\root\{name}\{version}".
+        # Also, if we allow abspaths, we probably want to prevent Windows
+        # abspaths on Linux (and vice versa).
+        raise ValueError(f"Input format string appears to be an absolute path: {format_string}")
+
+    # If we want to think of a string like "a/b/c" as a path (with 3 subdirs)
+    # on Windows, we cannot use pathlib (since "/" is not a path separator
+    # on Windows). Therefore, the path components are derived by splitting on
+    # both separators, on all operating systems.
+    components = re.split(any_sep, format_string)
```

Comment:
This will not work on all format strings. `{name}{/hash}` is a valid format string (the `/` sigil is included in the attribute for `hash` to ensure that it is colored as part of the hash rather than being colored separately). Because this would split in the middle of the attribute, it would fail to replace the hash.

Reply:
I have updated the function to forbid all components with `/`: from what I can tell, this is only for `{/hash}` and means that you want the literal string `"/aeflkjs..."` (which IMO should not be supported by `format_path`).